File: DesignPatterns/3-statePattern/typescript/state.ts
/**
 * -------------------------------------------------------
 * File: state.ts
 * Created Date: 2021-12-19 07:55:20
 * Author: ysj
 * Description: 状态类
 * -------------------------------------------------------
 */

import OrderContext from './context';

// 订单状态接口
export interface IOrderState {
  handle(ctx: OrderContext): void;
}

// 订单支付中
export class OrderPaying implements IOrderState {
  handle(ctx: OrderContext): void {
    console.log('订单支付中...');
    if (ctx.hasPaid) {
      ctx.setState(new OrderPaid());
      ctx.handle();
    } else if (ctx.elapsed30min) {
      ctx.setState(new OrderExpired());
      ctx.handle();
    } else {
      ctx.setState(new OrderUnpay());
      ctx.handle();
    }
  }
}

// 订单未支付
export class OrderUnpay implements IOrderState {
  handle(ctx: OrderContext): void {
    if (ctx.hasPaid) {
      ctx.setState(new OrderPaid());
      ctx.handle();
    } else if (ctx.elapsed30min) {
      ctx.setState(new OrderExpired());
      ctx.handle();
    } else {
      console.log('订单未支付...');
    }
  }
}

// 订单已支付
export class OrderPaid implements IOrderState {
  handle(ctx: OrderContext): void {
    console.log('订单已支付。');
  }
}

// 订单已过期
export class OrderExpired implements IOrderState {
  handle(ctx: OrderContext): void {
    console.log('订单已过期。');
  }
}

File: DesignPatterns/3-statePattern/python/state.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-
###
# File: state.py
# Created Date: 2021-12-19 06:12:44
# Author: ysj
# Description:  状态类
###

from abc import ABCMeta, abstractmethod
from context import OrderContext


class OrderState(metaclass=ABCMeta):
    """抽象基类"""
    @abstractmethod
    def handle(self, ctx: OrderContext):
        pass


class OrderPaying(OrderState):
    """订单支付中"""

    def handle(self, ctx: OrderContext):
        print("订单支付中...")
        if ctx.has_paid:
            ctx.set_state(OrderPaid())
            ctx.handle()
        elif ctx.elapsed_30min:
            ctx.set_state(OrderExpired())
            ctx.handle()
        else:
            ctx.set_state(OrderUnpay())
            ctx.handle()


class OrderUnpay(OrderState):
    """订单未支付"""

    def handle(self, ctx: OrderContext):
        if ctx.has_paid:
            ctx.set_state(OrderPaid())
            ctx.handle()
        elif ctx.elapsed_30min:
            ctx.set_state(OrderExpired())
            ctx.handle()
        else:
            print("订单未支付...")


class OrderPaid(OrderState):
    """订单已支付"""

    def handle(self, ctx: OrderContext):
        print("订单已支付。")


class OrderExpired(OrderState):
    """订单已过期"""

    def handle(self, ctx: OrderContext):
        print("订单已过期。")
